trust-quorum/tests/coordinator.rs

```diff
@@ -0,0 +1,646 @@
+    /// universe to select from, and we want to guarantee uniqueness. If we
+    /// derived `Ord` on `Index` and stored them in a BTreeSet, we'd get a
+    /// unique set of `Index` values. But when used, the output indexes can
+    /// still be duplicated due to the shift implementation used. Therefore we
+    /// instead just choose from a constrained set of usize values that we can
+    /// use directly as indexes into our fixed size structure for all tests.
+    #[strategy(btree_set(0..=255usize, MIN_CLUSTER_SIZE..=MAX_CLUSTER_SIZE))]
+    pub members: BTreeSet<usize>,
+
+    /// An index is roughly equivalent to a threshold, since a threshold cannot
+    /// exceed the number of members in a configuration. Using the index is
+    /// very efficient as opposed to a `prop_flat_map`. When we calculate the
+    /// threshold from the index we use max(2, Index), since the minimum
+    /// threshold is always 2.
+    pub threshold: Index,
+}
+
+/// All possible members used in a test
+fn member_universe() -> Vec<PlatformId> {
+    (0..=255)
+        .map(|serial| PlatformId::new("test".into(), serial.to_string()))
+        .collect()
+}
+
+const MIN_CLUSTER_SIZE: usize = 3;
+const MAX_CLUSTER_SIZE: usize = 32;
+const RETRY_TIMEOUT_MS: u64 = 100;
+
+#[derive(Debug, Arbitrary)]
+pub struct TestInput {
+    initial_config: GeneratedConfiguration,
+    actions: Vec<Action>,
+}
+
+#[proptest]
+fn test_coordinator_behavior_from_empty_state(input: TestInput) {
+    let logctx = test_setup_log("coordinator_behavior_from_empty_state");
+
+    let universe = member_universe();
+    let rack_id = RackUuid::new_v4();
+
+    // We don't know the coordinator ID until we translate the indexes in the
+    // test input to `PlatformId`s.
+    let coordinator_id = None;
+    let initial_config_msg = generated_config_to_reconfigure_msg(
+        &universe,
+        coordinator_id,
+        Epoch(1),
+        rack_id,
+        input.initial_config,
+    );
+
+    let coordinator_id = initial_config_msg.members.first().unwrap().clone();
+    let mut state = TestState::new(logctx.log.clone(), coordinator_id);
+
+    // Start the initial configuration
+    state.action_coordinate_reconfiguration(initial_config_msg)?;
+
+    // Start executing the actions
+    state.run_actions(input.actions)?;
+
+    logctx.cleanup_successful();
+}
+
+fn generated_config_to_reconfigure_msg(
+    member_universe: &[PlatformId],
+    coordinator_id: Option<&PlatformId>,
+    epoch: Epoch,
+    rack_id: RackUuid,
+    config: GeneratedConfiguration,
+) -> ReconfigureMsg {
+    let mut members: BTreeSet<PlatformId> = config
+        .members
+        .iter()
+        .map(|index| member_universe[*index].clone())
+        .collect();
+
+    if let Some(coordinator_id) = coordinator_id {
+        members.insert(coordinator_id.clone());
+    }
+
+    println!("members = {members:#?}");
+
+    let threshold =
+        Threshold(usize::max(2, config.threshold.index(members.len())) as u8);
```

Comment:
Hmm I thought `index.index` panicked if the value passed in was 0.

Reply:
It does, but we never generate a set of members smaller than 3 (`MIN_CLUSTER_SIZE`).

Reply:
To clarify, we just don't want the index to be 0 or 1. The length of the set is never smaller than 3.